// websocket-framed/lib/websocket-framed.node.js
"use strict";

var _eventemitter = _interopRequireDefault(require("eventemitter3"));

var _encodr = _interopRequireDefault(require("encodr"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
**  WebSocket-Framed -- Framed WebSocket Communication
**  Copyright (c) 2017-2019 Dr. Ralf S. Engelschall <rse@engelschall.com>
**
**  Permission is hereby granted, free of charge, to any person obtaining
**  a copy of this software and associated documentation files (the
**  "Software"), to deal in the Software without restriction, including
**  without limitation the rights to use, copy, modify, merge, publish,
**  distribute, sublicense, and/or sell copies of the Software, and to
**  permit persons to whom the Software is furnished to do so, subject to
**  the following conditions:
**
**  The above copyright notice and this permission notice shall be included
**  in all copies or substantial portions of the Software.
**
**  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
**  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
**  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
**  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
**  CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
**  TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
**  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*  external dependencies  */

/*  helper function: define a non-enumerable property  */
const mkProp = (obj, name, value, writeable = true) => {
  Object.defineProperty(obj, name, {
    configurable: false,
    enumerable: false,
    writable: writeable,
    value: value
  });
};
/*  internal class: a single data frame  */


class Frame {
  constructor(channel) {
    mkProp(this, "channel", channel);
    this.reset();
  }

  reset() {
    this.fid = 0;
    this.rid = 0;
    this.type = "MESSAGE";
    this.data = null;
  }

  set(frame, replyTo = undefined) {
    this.fid = frame.fid ? frame.fid : this.channel.fid();
    this.rid = frame.rid ? frame.rid : replyTo ? replyTo.fid : 0;
    this.type = frame.type ? frame.type : "MESSAGE";
    this.data = frame.data ? frame.data : null;
    return this;
  }

  import(data) {
    let frame = this.channel.codec.decode(data);
    this.fid = frame[0];
    this.rid = frame[1];
    this.type = frame[2];
    this.data = frame[3];
    return this;
  }

  export() {
    let frame = [this.fid, this.rid, this.type, this.data];
    let data = this.channel.codec.encode(frame);
    return data;
  }

}
/*  API class: framed WebSocket communication  */


class WebSocketFramed extends _eventemitter.default {
  constructor(ws, codec = "json") {
    super();
    mkProp(this, "ws", ws);
    mkProp(this, "codec", new _encodr.default(codec));
    mkProp(this, "fid", (() => {
      let fid = 0;
      return () => {
        fid++;
        if (fid >= 0xffffffff - 1) fid = 1;
        return fid;
      };
    })());
    this.ws.addEventListener("message", ev => {
      let frame = new Frame(this);

      try {
        let data = ev.data;
        if (typeof data !== "string" && !ArrayBuffer.isView(data) && process.browser) data = new Uint8Array(data);
        frame.import(data);
      } catch (ex) {
        this.emit("error", ex);
        return;
      }

      ev.frame = frame;
      this.emit("message", ev);
    });
  }

  send(frameData, replyTo) {
    let frame = new Frame(this);
    frame.set(frameData, replyTo);
    let data = frame.export();
    if (typeof data !== "string" && ArrayBuffer.isView(data) && process.browser) data = data.buffer;
    this.ws.send(data);
    return {
      frame,
      data
    };
  }

}
/*  export the API class the traditional way  */


module.exports = WebSocketFramed;
